acceptance_modelisation/base_acceptance_map_creator.py

```diff
@@ -456,62 +460,67 @@ def _create_model_cos_zenith_binned(self,
+            # Create an array of cumulative weight of the selected variable vs cos(zenith)
+            cumsum_variable[wobble] = np.cumsum(np.histogram(cos_zenith_observations[wobble_observations == wobble],
+                                                             bins=cos_zenith_bin,
+                                                             weights=cut_variable_weights[
+                                                                 wobble_observations == wobble])[0])
+        # Initiate the list of index of selected zenith bin edges
+        zenith_selected = [0]
 
-        # Adapt binning to have a given number of minimum livetime or run in each bin
         i = 0
-        at_least_2_wobble, cut_variable_min_for_each_wobble, condition_is_fulfilled_per_wobble = False, False, False
-
-        while i < len(cut_variable_per_bin):
-            is_last_bin = (i + 1) == len(cut_variable_per_bin)
-            in_coszd_bin = (cos_zenith_observations >= cos_zenith_bin[i]) & (
-                    cos_zenith_observations < cos_zenith_bin[i + 1])
-
-            if per_wobble:
-                wobble_in_bin = wobble_observations[in_coszd_bin]
-                cut_variable_in_bin = cut_variable_weights[in_coszd_bin]
-                cut_variable_in_bin_per_wobble = np.array(
-                    [cut_variable_in_bin[wobble_in_bin == wobble].sum() for wobble in np.unique(wobble_in_bin)])
-                at_least_2_wobble = len(np.unique(wobble_in_bin)) > 1
-                if at_least_2_wobble:
-                    cut_variable_min_for_each_wobble = np.all(
-                        cut_variable_in_bin_per_wobble >= min_cut_per_cos_zenith_bin)
-                condition_is_fulfilled_per_wobble = at_least_2_wobble and cut_variable_min_for_each_wobble
-
-            condition_is_fulfilled_in_bin = cut_variable_per_bin[i] >= min_cut_per_cos_zenith_bin
-
-            if ((per_wobble and not condition_is_fulfilled_per_wobble) or (not condition_is_fulfilled_in_bin)):
-                if not is_last_bin:
-                    cut_variable_per_bin[i] += cut_variable_per_bin[i + 1]
-                    cut_variable_per_bin = np.delete(cut_variable_per_bin, i + 1)
-                    cos_zenith_bin = np.delete(cos_zenith_bin, i + 1)
-                
-                elif i > 0:
-                    cut_variable_per_bin[i - 1] += cut_variable_per_bin[i]
-                    cut_variable_per_bin = np.delete(cut_variable_per_bin, i)
-                    cos_zenith_bin = np.delete(cos_zenith_bin, i)
-                    i -= 1
-                else:
-                    i += 1
+        n = len(cos_zenith_bin) - 2
+
+        while i < n:
+            # For each wobble, find the index of the first zenith which fulfills the zd binning criteria if any
+            # Then flatten and sort the array
+            candidate_i = np.sort((np.concatenate([np.ravel(
+                np.argwhere(cum_cut_variable >= self.cos_zenith_binning_parameter_value))[:1]
```

Comment:
From numpy documention, it seem they are no guarantee than the output of argwhere is ordered. Could be good to add a sort here to ensure no bugs appear in the future.

Reply:
`numpy.nonzero` documentation indicates that  "The values in 'the input array' are always tested and returned in row-major, C-style order.". And this output is directly given to argwhere (after transposing).
And I don't see  a reason for it to change to random order in the future.

Still, we could add a sort, or unit tests to check the behaviour. I would prefer the later as there is already a lot of manipulation done here and I would prefer not including effectless function calls. What do you think?

Reply:
Let's try to keep in mind this point, maybe with a comment in the code precising we assume it to be ordered.

Reply:
I replaced argwhere with nonzero. After testing argwhere was adding a dimension during tranposed thus requiring to remove it with ravel afterward. Both operations are skipped by using nonzero.

Adding the comment about nonzero being assumed ordered.